File: frontend/src/app/interfaces/models/response-model.ts

export type ResponseModel<T> = { data: T } | { error: string };

export function responseIsError<T>(response: ResponseModel<T>): response is { error: string } {
    return (response as any).error;
}

export function responseIsOk<T>(response: ResponseModel<T>): response is { data: T } {
    return (response as any).data;
}

File: backend/src/main/java/org/acme/models/ResponseModel.java
package org.acme.models;

public class ResponseModel<T> {
    private T data;
    private String error;

    public ResponseModel(T data) {
        this.data = data;
    }

    public ResponseModel(String error) {
        this.error = error;
    }

    public T getData() {
        return data;
    }

    public void setData(T data) {
        this.data = data;
    }

    public String getError() {
        return error;
    }

    public void setError(String error) {
        this.error = error;
    }
}
